### PR TITLE
[PSM Interop] Initial support for GAMMA tests

### PR DESCRIPTION
**Pull request by sergiitk**
Adds initial support for K8s [GAMMA](https://gateway-api.sigs.k8s.io/concepts/gamma/) (Gateway API for Service Mesh) initiative.

- Add framework support for loading CRD-based APIs using k8s python dynamic client
- Add basic mesh baseline test (aka ping-pong) using GAMMA setup
- Implement initial framework changes needed to run PSM tests on GAMMA-enabled cluster using [TDMesh](https://cloud.google.com/traffic-director/docs/gke-gateway-overview#gateway-api) and GRPCRoute.

Based on https://github.com/grpc/grpc/pull/33504.

### CODE REVIEW

**Comment by sergiitk** [comment]
- [ ] [grpc/java/v1.57.x/branch/xds_k8s_lb](https://source.cloud.google.com/results/invocations/7786bb71-ae53-44af-b1cd-84a6cf90ada1)
- [ ] [grpc/java/v1.57.x/branch/psm-security](https://source.cloud.google.com/results/invocations/14d09c96-1114-400d-9cb8-6a90a75285c0)
- [ ] [grpc/java/v1.57.x/branch/xds_url_map](https://source.cloud.google.com/results/invocations/78159abc-0d61-4a68-a63c-f3ff24c8257b)

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director_gamma.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import framework.infrastructure.traffic_director as td_base
+
+
+class TrafficDirectorGammaManager(td_base.TrafficDirectorManager):
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director_gamma.py:17]: TODO: make a TD-manager-less base test case

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_base_runner.py [hunk 2]
```diff
@@ -250,9 +250,15 @@ def _template_file_from_name(cls, template_name):
...
+    def _create_from_template(
+        self,
+        template_name,
+        *,
+        custom_object: bool = False,
+        **kwargs,
+    ) -> object:
         template_file = self._template_file_from_name(template_name)
-        logger.debug("Loading k8s manifest template: %s", template_file)
+        logger.info("Loading k8s manifest template: %s", template_file)
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_base_runner.py:261]: todo: undo

File: tools/run_tests/xds_k8s_test_driver/framework/xds_gamma_testcase.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+GammaServerRunner = gamma_server_runner.GammaServerRunner
+KubernetesClientRunner = k8s_xds_client_runner.KubernetesClientRunner
+XdsTestClient = client_app.XdsTestClient
+XdsTestServer = server_app.XdsTestServer
+
+logger = logging.getLogger(__name__)
+
+
+class GammaXdsKubernetesTestCase(xds_k8s_testcase.RegularXdsKubernetesTestCase):
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_gamma_testcase.py:33]: todo: move into `framework/test_cases`

File: tools/run_tests/xds_k8s_test_driver/framework/xds_gamma_testcase.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    def setUp(self):
+        """Hook method for setting up the test fixture before exercising it."""
+        # Skip a class.
+        super(xds_k8s_testcase.IsolatedXdsKubernetesTestCase, self).setUp()
+
+        # Random suffix per test.
+        self.createRandomSuffix()
+
+        # TD Manager
+        self.td = self.initTrafficDirectorManager()
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_gamma_testcase.py:47]: TODO: make a TD-manager-less base test case

File: tools/run_tests/xds_k8s_test_driver/framework/xds_gamma_testcase.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        self.client_namespace = KubernetesClientRunner.make_namespace_name(
+            self.resource_prefix, self.resource_suffix
+        )
+        self.client_runner = self.initKubernetesClientRunner()
+
+        # Cleanup.
+        self.force_cleanup = True
+        self.force_cleanup_namespace = True
+
+    def initTrafficDirectorManager(
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_gamma_testcase.py:73]: TODO: make a TD-manager-less base test case

File: tools/run_tests/xds_k8s_test_driver/tests/gamma/gamma_baseline_test.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+flags.adopt_module_key_flags(xds_k8s_testcase)
+
+_XdsTestServer = xds_k8s_testcase.XdsTestServer
+_XdsTestClient = xds_k8s_testcase.XdsTestClient
+
+
+class GammaBaselineTest(xds_gamma_testcase.GammaXdsKubernetesTestCase):
+    def test_ping_pong(self):
+        with self.subTest("1_run_test_server"):
+            test_server: _XdsTestServer = self.startTestServers()[0]
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/tests/gamma/gamma_baseline_test.py:32]: TODO: Consider moving out custom gamma resource creation out of `self.startTestServers()`

File: tools/run_tests/xds_k8s_test_driver/bin/run_test_server.py [hunk 7]
```diff
@@ -28,8 +34,11 @@
 _CMD = flags.DEFINE_enum(
     "cmd", default="run", enum_values=["run", "cleanup"], help="Command"
 )
-_SECURE = flags.DEFINE_bool(
```
Comment (gnossen) [inline tools/run_tests/xds_k8s_test_driver/bin/run_test_server.py:31]: Was this flag used anywhere?

**Review by gnossen — APPROVED** [review]
Looks great! Thanks for making this happen.

File: tools/run_tests/xds_k8s_test_driver/kubernetes-manifests/gamma/tdmesh.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  labels:
+    owner: xds-k8s-interop-test
+spec:
+  gatewayClassName: gke-td
+  allowedRoutes:
+    namespaces:
+      from: All
+    kinds:
+    - group: net.gke.io
+      kind: TDGRPCRoute
```
[the remark this answers is not in the document]
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/kubernetes-manifests/gamma/tdmesh.yaml:16]: Confirmed this doesn't work at the moment:

```
framework.helpers.retryers.RetryError: Retry error calling framework.infrastructure.k8s.KubernetesNamespace.get_service: timeout 0:01:00 (h:mm:ss) exceeded. Check result callback returned False.

Reason: A k8s service wasn't assigned a NEG (Network Endpoint Group).
################################################################################
# IMPORTANT: This is not a root cause. This is an indication that
# _something_ -- literally _anything_ -- has gone wrong in the xDS flow.
# It is _your_ responsibility to look through the interop client and/or
# server logs to determine what exactly went wrong.
################################################################################
# Please inspect the information below:
Timeout 0:01:00 (h:mm:ss) waiting for service psm-grpc-server to report NEG status. Last service status:
psm-grpc-server:
---
conditions: null
load_balancer:
  ingress: null
...
```

**Comment by sergiitk** [comment]
- [ ] [grpc/java/v1.57.x/branch/xds_k8s_lb](https://source.cloud.google.com/results/invocations/55a63156-93a1-4bbb-a52b-99a70346985e)
- [ ] [grpc/java/v1.57.x/branch/psm-security](https://source.cloud.google.com/results/invocations/1e9d8504-f2d8-4978-ad2e-42a69f0be769)
- [ ] [grpc/java/v1.57.x/branch/xds_url_map](https://source.cloud.google.com/results/invocations/5ff387dd-7fc7-42b5-bc68-7d3107784a50)

**Pull request merged**